src/javascript/crypto/e2e/coname/lookup_verify.js

```
@@ -0,0 +1,602 @@
+ * @param {Array.<boolean>} prefixBits The prefix bits
+ * @param {object} node The ReconstructedNode
+ * @return {!e2e.ByteArray}
+ */
+e2e.coname.recomputeHash_ = function(treeNonce, prefixBits, node) {
+  var shake256 = e2e.coname.sha3.shake256(e2e.coname.MERKLE_HASH_BYTES);
+  if (!node) {
+    // return HashEmptyBranch(treeNonce, prefixBits);
+    // This is the same as in the CONIKS paper.
+    // H(k_empty || nonce || prefix || depth)
+    return shake256.update(e2e.coname.MERKLE_NODEID_EMPTY_BRANCH).
+        update(treeNonce).
+        update(e2e.coname.toBytes_(prefixBits)).
+        update(e2e.coname.numberTo4ByteArray_(prefixBits.length)).
+        digest();
+
+  } else if (!node.children) { // i.e., isLeaf()
+    // return HashLeaf(treeNonce, node.index, node.depth, node.value);
+    // This is the same as in the CONIKS paper:
+    // H(k_leaf || nonce || index || depth || value)
+    return shake256.update(e2e.coname.MERKLE_NODEID_LEAF).
+        update(treeNonce).
+        update(node.index).
+        update(e2e.coname.numberTo4ByteArray_(node.depth)).
+        update(node.value).
+        digest();
+
+  } else {
+    for (var h, childHashes = [], i = 0; i < 2; i++) {
+      h = node.children[i];
+
+      childHashes[i] = h.Omitted ?
+          h.Omitted :
+          e2e.coname.recomputeHash_(treeNonce,
+          prefixBits.concat(i === 1),
+          h.Present);
+    }
+
+    // return HashInternalNode(prefixBits, childHashes);
+    // Differences from the CONIKS paper:
+    // * Add an identifier byte at the beginning to make it impossible for this
+    //   to collide with leaves or empty branches.
+    // * Add the prefix of the index, to protect against limited hash
+    //   collisions or bugs.
+    // This gives H(k_internal || h_child0 || h_child1 || prefix || depth)
+    return shake256.
+        update(e2e.coname.MERKLE_NODEID_INTERNAL).
+        update(childHashes[0] || []).
+        update(childHashes[1] || []).
```

`childHashes[0]` and `childHashes[1]` should never be null/undefined, right? So I don't think we want the `|| []` -- it's better to throw an error (actually, would `update` throw an error?) than to fail silently.